src/Microsoft.AspNet.Mvc.TagHelpers/CacheTagHelper.cs

```diff
@@ -0,0 +1,262 @@
+        /// </summary>
+        [HtmlAttributeName(ExpiresSlidingAttributeName)]
+        public TimeSpan? ExpiresSliding { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <see cref="CachePreservationPriority"/> policy for the cache entry.
+        /// </summary>
+        [HtmlAttributeName(CachePriorityAttributeName)]
+        public CachePreservationPriority? Priority { get; set; }
+
+        /// <inheritdoc />
+        public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
+        {
+            var key = GenerateKey(context);
+            string result;
+            if (!MemoryCache.TryGetValue(key, out result))
+            {
+                result = await context.GetChildContentAsync();
+                MemoryCache.Set(key, cacheSetContext =>
+                {
+                    UpdateCacheContext(cacheSetContext);
+                    return result;
+                });
+            }
+
+            // Clear the contents of the "cache" element since we don't want to render it.
+            output.SupressOutput();
+
+            output.Content = result;
+        }
+
+        // Internal for unit testing
+        internal string GenerateKey(TagHelperContext context)
+        {
+            var builder = new StringBuilder(nameof(CacheTagHelper));
+            builder.Append(CacheKeyTokenSeparator)
+                   .Append(context.UniqueId);
+
+            var request = ViewContext.HttpContext.Request;
+
+            if (!string.IsNullOrEmpty(VaryBy))
+            {
+                builder.Append(CacheKeyTokenSeparator)
+                       .Append(nameof(VaryBy))
+                       .Append(CacheKeyTokenSeparator)
+                       .Append(VaryBy);
+            }
+
+            AddStringCollectionKey(builder, nameof(VaryByCookie), VaryByCookie, request.Cookies);
+            AddStringCollectionKey(builder, nameof(VaryByHeader), VaryByHeader, request.Headers);
+            AddStringCollectionKey(builder, nameof(VaryByQuery), VaryByQuery, request.Query);
+            AddVaryByRouteKey(builder);
+
+            if (VaryByUser)
+            {
+                builder.Append(CacheKeyTokenSeparator)
+                       .Append(nameof(VaryByUser))
+                       .Append(CacheKeyTokenSeparator)
+                       .Append(ViewContext.HttpContext.User?.Identity?.Name);
+            }
+
+            // The key is typically too long to be useful, so we use a cryptographic hash
+            // as the actual key (better randomization and key distribution, so small vary
+            // values will generate dramtically different keys).
+            using (var sha = SHA256.Create())
+            {
+                var contentBytes = Encoding.UTF8.GetBytes(builder.ToString());
+                var hashedBytes = sha.ComputeHash(contentBytes);
+                return Convert.ToBase64String(hashedBytes);
```

Comment:
use streaming overloads if possible instead of tostring and arrays, as the will cause LOH hits.